// botbuilder-ai/lib/qnamaker-utils/trainUtils.js
"use strict";
/**
 * @module botbuilder-ai
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const httpRequestUtils_1 = require("./httpRequestUtils");
/**
 * Generate Answer api utils class.
 *
 * @remarks
 * This class is helper class for generate answer api, which is used to make queries to a single QnA Maker knowledge base and return the result.
 */
class TrainUtils {
    /**
     * Creates new instance for active learning train utils.
     * @param endpoint The endpoint of the knowledge base to query.
     */
    constructor(endpoint) {
        this.endpoint = endpoint;
        this.httpRequestUtils = new httpRequestUtils_1.HttpRequestUtils();
    }
    /**
     * Train API to provide feedback.
     * @param feedbackRecords Feedback record list.
     */
    async callTrainAsync(feedbackRecords) {
        if (!feedbackRecords) {
            throw new TypeError('Feedback records can not be null.');
        }
        if (!feedbackRecords.feedbackRecords || feedbackRecords.feedbackRecords.length == 0) {
            return;
        }
        await this.queryTrainAsync(feedbackRecords);
    }
    async queryTrainAsync(feedbackRecords) {
        const url = `${this.endpoint.host}/knowledgebases/${this.endpoint.knowledgeBaseId}/train`;
        var payloadBody = JSON.stringify({
            feedbackRecords: feedbackRecords
        });
        await this.httpRequestUtils.executeHttpRequest(url, payloadBody, this.endpoint);
    }
}
exports.TrainUtils = TrainUtils;
